craft_parts/state_manager/step_state.py

```diff
@@ -0,0 +1,121 @@
+    """
+
+    part_properties: Dict[str, Any] = Field({}, alias="properties")
+    project_options: Dict[str, Any] = {}
+    files: Set[str] = set()
+    directories: Set[str] = set()
+
+    class Config:
+        """Pydantic model configuration."""
+
+        validate_assignment = True
+        extra = "ignore"
+        allow_mutation = False
+        alias_generator = lambda s: s.replace("_", "-")  # noqa: E731
+        allow_population_by_field_name = True
+
+    @abstractmethod
+    def properties_of_interest(self, part_properties: Dict[str, Any]) -> Dict[str, Any]:
+        """Return relevant properties concerning this step."""
+
+    @abstractmethod
+    def project_options_of_interest(
+        self, project_options: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        """Return relevant project options concerning this step."""
+
+    def diff_properties_of_interest(self, other_properties: Dict[str, Any]) -> Set[str]:
+        """Return properties of interest that differ.
+
+        Take a dictionary of properties and compare to our own, returning
+        the set of property names that are different. Both dictionaries
+        are filtered prior to comparison, only relevant properties are
+        compared.
+
+        :param other_properties: The properties to compare to the
+            project options stored in this state.
+        """
+        return _get_differing_keys(
+            self.properties_of_interest(self.part_properties),
+            self.properties_of_interest(other_properties),
+        )
+
+    def diff_project_options_of_interest(
+        self, other_project_options: Dict[str, Any]
+    ) -> Set[str]:
+        """Return project options that differ.
+
+        Take a dictionary of project_options and compare to our own,
+        returning the set of project option names that are different. Both
+        dictionaries are filtered prior to comparison, only relevant
+        options are compared.
+
+        :param other_project_options: The project options to compare to
+           the project options stored in this state.
+        """
+        return _get_differing_keys(
+            self.project_options_of_interest(self.project_options),
+            self.project_options_of_interest(other_project_options),
+        )
+
+    def marshal(self) -> Dict[str, Any]:
+        """Create a dictionary containing the part state data.
+
+        :return: The newly created dictionary.
+        """
+        return self.dict(by_alias=True)
+
+    def write(self, filepath: Path) -> None:
+        """Write this state to disk."""
+        os.makedirs(filepath.parent, exist_ok=True)
+        yaml_data = self.yaml(by_alias=True)
+        with open(filepath, "w") as f:
+            f.write(yaml_data)
+
+
+def _get_differing_keys(dict1, dict2) -> Set[str]:
+    differing_keys = set()
+    for key, dict1_value in dict1.items():
+        dict2_value = dict2.get(key)
+        if dict1_value != dict2_value:
+            differing_keys.add(key)
```

Comment:
For the following case, should `foobar` be included in the `differing_keys` set?

- dict1: `{'foobar': None}`
- dict2: `{}`

Reply:
It shouldn't, will document the rules.

Reply:
_get_differing_keys can be simplified to:

diff_items = (dict1.items() - dict2.items()) | (dict2.items() - dict1.items())`
return [i[0] for i in diff_items]

Reply:
Only if all the values from both dicts are hashable. Signature type hints say values can be `Any`.